compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -602,4 +612,56 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         }
         self.flounder(&responses)
     }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    pub(super) fn evaluate_alias_bound_self_is_well_formed<G: GoalKind<'tcx>>(
+        &mut self,
+        goal: Goal<'tcx, G>,
+    ) -> QueryResult<'tcx> {
+        match *goal.predicate.self_ty().kind() {
+            ty::Alias(ty::Projection, projection_ty) => {
+                let mut param_env_candidates = vec![];
+                let projection_trait_ref = projection_ty.trait_ref(self.tcx());
+
+                if projection_trait_ref.self_ty().is_ty_var() {
+                    return self
+                        .evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS);
+                }
+
+                let trait_goal = goal.with(
+                    self.tcx(),
+                    ty::TraitPredicate {
+                        trait_ref: projection_trait_ref,
+                        constness: ty::BoundConstness::NotConst,
+                        polarity: ty::ImplPolarity::Positive,
+                    },
+                );
+                self.assemble_param_env_candidates(trait_goal, &mut param_env_candidates);
+
+                // FIXME: Bad bad bad bad bad !!!!!
+                let lang_items = self.tcx().lang_items();
+                let trait_def_id = projection_trait_ref.def_id;
+                let funky_built_in_res = if lang_items.pointee_trait() == Some(trait_def_id) {
+                    G::consider_builtin_pointee_candidate(self, goal)
+                } else if lang_items.discriminant_kind_trait() == Some(trait_def_id) {
+                    G::consider_builtin_discriminant_kind_candidate(self, goal)
+                } else {
+                    Err(NoSolution)
+                };
+                if let Ok(result) = funky_built_in_res {
+                    param_env_candidates
+                        .push(Candidate { source: CandidateSource::BuiltinImpl, result });
+                }
```

Comment:
See https://hackmd.io/frHMYxD5S4-rWlxkx_eWNw#1-The-curious-issue-with-DiscriminantKind